File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(strings, predicate, /) -> str:
    for string in strings: #For each string, test the predicate and return the lowercase version of the first string that passes
        if predicate(string):
            return string.lower()
    return None #Otherwise, return None

# Write your powers generator here
def powers_generator(*, base, limit) -> int:
    curr_number = 1
    while curr_number <= limit: #While the current number is below our limit, keep yielding it and multiplying it by the base
        yield curr_number
        curr_number = curr_number * base

# Write your say function here
def say(word = None, /) -> str:
    if word == None: #Check base case
        return ""
    else: #Otherwise, return function that concatenates subsequent calls and recurses
        def say_next(next_word = None, /) -> str:
            if next_word == None: #Check base case
                return word
            else:
                return say(word + " " + next_word)
    return say_next


# Write your line count function here
def meaningful_line_count(filepath) -> int:
    meaningful_lines = 0
    with open(filepath) as file: #Open file
        for line in file: #For each line in the file, remove the brightspace and check if it's empty or starts with "#"
            filtered_line = ''.join(line.split())
            if filtered_line != "" and filtered_line[0] != "#":
                meaningful_lines += 1
        file.close() #Close the file when we're done
    return meaningful_lines


# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion():
    def __init__(self, a, b, c, d) -> object:
        object.__setattr__(self, "a", a)
        object.__setattr__(self, "b", b)
        object.__setattr__(self, "c", c)
        object.__setattr__(self, "d", d)

    def __add__(self, other) -> object:
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
    
    def __mul__(self, other) -> object:
        # Hamilton product formula taken from: https://en.wikipedia.org/wiki/Quaternion#Hamilton_product 
        return Quaternion(((self.a * other.a) - (self.b * other.b) - (self.c * other.c) - (self.d * other.d)), 
                          ((self.a * other.b) + (self.b * other.a) + (self.c * other.d) - (self.d * other.c)), 
                          ((self.a * other.c) - (self.b * other.d) + (self.c * other.a) + (self.d * other.b)), 
                          ((self.a * other.d) + (self.b * other.c) - (self.c * other.b) + (self.d * other.a)))
    
    @property
    def coefficients(self) -> tuple:
        return (self.a, self.b, self.c, self.d)
    
    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __eq__(self, other) -> bool:
        return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d

    def __str__(self) -> str:
        str_representation = ""
        my_coefficients = self.coefficients
        for index, value in enumerate(my_coefficients): #For each of our values, we need to respond differently based on the values
            if value != 0: #If any of our values are 0, we don't put them in the string (we deal with the empty Quaternion case later)
                if str_representation != "" and value > 0: #If the value is positive and the string isn't empty, add the addition sign
                    str_representation += "+"
                elif value < 0: #If the value is negative, add the negative/subtraction sign
                    str_representation += "-"
                if (value != 1 and value != -1) or index == 0: #If the value isn't +-1 or it's the value with no variable, add the value
                    str_representation += str(abs(value))
                match index: #Depending on the index, add the corresponding variable
                    case 1:
                        str_representation += "i"
                    case 2:
                        str_representation += "j"
                    case 3: 
                        str_representation += "k"
        
        if str_representation == "": # Deal with edge case of an empty quaternion
            str_representation = "0"
        return str_representation

File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

import java.lang.Math;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    // Write your first then lower case function here
    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream() //Grab stream of strings so we can perform some aggregate operations
                .filter(predicate)  //Filter for the predicate
                .findFirst()    //Grab first
                .map(String::toLowerCase);  //Make that string lowercase
    }

    // Write your say function here
    ///Speaker class for the say method
    static class Speaker{
        //Phrase of the current speaker
        private String phrase;

        //Constructor
        public Speaker(String initialPhrase){
            this.phrase = initialPhrase;
        }

        //And method -> returns a new Speaker with the extended phrase
        public Speaker and(String word){
            return new Speaker(this.phrase + " " + word);
        }

        //Phrase getter method
        public String phrase(){
            return this.phrase;
        }
    }

    //Say method for empty argument case
    public static Speaker say(){
        return new Speaker("");
    }

    //Say method for String argument case
    public static Speaker say(String word){
        return new Speaker(word);
    }

    // Write your line count function here
    public static int meaningfulLineCount(String path) throws IOException{
        try (var reader = new BufferedReader(new FileReader(path))) {
            return (int) reader.lines()
                            .filter(line -> !line.trim().isBlank() && !line.trim().startsWith("#"))
                            .count();
        }
    }
}

// Write your Quaternion record class here
record Quaternion(double a, double b, double c, double d) {
    public Quaternion{
        if(a != a || b != b || c != c || d != d){
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }
    //Static pre-made Quaternions
    public final static Quaternion ZERO = new Quaternion(0.0, 0.0, 0.0, 0.0);
    public final static Quaternion I = new Quaternion(0.0, 1.0, 0.0, 0.0);
    public final static Quaternion J = new Quaternion(0.0, 0.0, 1.0, 0.0);
    public final static Quaternion K = new Quaternion(0.0, 0.0, 0.0, 1.0);

    Quaternion plus(Quaternion other){
        return new Quaternion(a + other.a, b + other.b, c + other.c, d + other.d);
    }

    Quaternion times(Quaternion other){
        return new Quaternion(
                            ((a * other.a)-(b * other.b)-(c * other.c)-(d * other.d)),
                            ((a * other.b)+(b * other.a)+(c * other.d)-(d * other.c)),
                            ((a * other.c)-(b * other.d)+(c * other.a)+(d * other.b)),
                            ((a * other.d)+(b * other.c)-(c * other.b)+(d * other.a))
                            );
    }

    public Quaternion conjugate(){
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients(){
        return List.of(a, b, c, d);
    }

    @Override
    public String toString(){
        //Zero case
        if(this.coefficients().equals(ZERO.coefficients())){
            return "0";
        }
        //Non-zero case
        String strRepresentation = "";
        List<Double> myCoefficients = coefficients();
        for (int i = 0; i < myCoefficients.size(); i ++){
            double coefficient = myCoefficients.get(i);
            //System.out.print("\n" + coefficient);
            if (coefficient != 0.0){
                strRepresentation += (coefficient > 0.0 && !strRepresentation.isBlank()) ? "+" : "";
                strRepresentation += (coefficient < 0.0) ? "-" : "";
                strRepresentation += ((Math.abs(coefficient) != 1) || i == 0) ? Math.abs(coefficient) : "";
                switch (i){
                    case 1:
                        strRepresentation += "i";
                        break;
                    case 2:
                        strRepresentation += "j";
                        break;
                    case 3:
                        strRepresentation += "k";
                        break;
                    default:
                        break;
                }
            }
        }
        //System.out.print("\n" + strRepresentation);
        return strRepresentation;
    }
    
}


// Write your BinarySearchTree sealed interface and its implementations here
sealed interface BinarySearchTree permits Empty, Node{
    int size();

    boolean contains(String value);

    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree{
    @Override
    public int size(){
        return 0;
    }

    @Override
    public boolean contains(String value){
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString(){
        return "()";
    }
}

final class Node implements BinarySearchTree{
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0){
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        String BSTString = "(" + (left.size() > 0 ? left + value: value);
        BSTString += right.size() > 0 ? right + ")": ")";
        return BSTString;
    }
}